### DISCUSSION TITLE
How to set Ocelot HttpConnection timeout?

### DISCUSSION THREAD

**Discussion by SamYuann** [ThreeMammals/Ocelot#1832] [Q&A] [answered]
I set QoSOptions TimeoutValue 60s:
```json
"QoSOptions": {
        "ExceptionsAllowedBeforeBreaking": 3,
        "DurationOfBreak": 5,
        "TimeoutValue": 60000
}
```
And my response complete with 33.78s, but web client catch “gateway timeout”，this is ocelot log:
```
2019-01-02 09:14:25.3196|2|INFO|Microsoft.AspNetCore.Hosting.Internal.WebHost|Request finished in 34134.0082ms 200 application/vnd.openxmlformats-officedocument.spreadsheetml.sheet
2019-01-02 09:14:25.3208|2|DEBUG|Microsoft.AspNetCore.Server.Kestrel|Connection id "0HLJBPGDC9PUH" stopped.
```
I think this is `HttpConnection` Timeout, but how to set Ocelot `HttpConnection`'s Timeout?

**Comment by ks1990cn** [comment]
@SamYuann  I don't get your question. Just clarify I understood can you explain:
 Response complete with 33.78s (while calling through gateway or directly downstream).

If its by gateway call, and response returns then Ocelot is fine, check with webclient and next thing you are trying to do with.

Exactly I don't remember solution for your problem but few years ago I have seen this problem in another project (Not in Ocelot),
and I searched on stackoverflow at that time to fix it,

Probably something wrong with http or https scheme.
https://stackoverflow.com/questions/58547548/dotnet-run-on-diferent-port-causes-a-connection-id-xxxx-request-processing-en
<br/>
https://github.com/dotnet/aspnetcore/issues/44870

I need explanation again to understand problem, I haven't seen these things with latest Ocelot version and I have done testings over it.

Are you still using Ocelot? @SamYuann

**Comment by raman-m** [comment]
@ks1990cn 
Mr. Sam is not even with GitHub )) [Repos are empty](https://github.com/SamYuann?tab=repositories)! Also, activities graph is empty since 2018/19. There is no forked Ocelot repo...

**Conclusion**: `The next ghost is detected!`

**Comment by raman-m** [comment, marked as answer]
> @SamYuann commented [on Jan 2, 2019](https://github.com/ThreeMammals/Ocelot/discussions/1832#discussion-5923837)

Sam, welcome to Ocelot world!

-----
> I think this is `HttpConnection` Timeout

Why do you think so? I don't see any problems at all! Your request was finished in 34.134 seconds with successful status code `200` aka `OK`. What's the problem, my dear? 😉 
So, QoS Timeout policy wasn't applied (34 is less than 60). 😄 
And default `HttpClient` timeout is **90** secs:
- `_defaultTimeout = TimeSpan.FromSeconds(90);`
- Code link: [Ocelot.Requester.HttpClientBuilder constructor initializes def timeout by 90 value](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Requester/HttpClientBuilder.cs#L27)

And if QoS value is defined then it is used instead of default 90 secs value:
```csharp
            var timeout = downstreamRoute.QosOptions.TimeoutValue == 0
                ? _defaultTimeout
                : TimeSpan.FromMilliseconds(downstreamRoute.QosOptions.TimeoutValue);
```
See the Time out logic code [here](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Requester/HttpClientBuilder.cs#L53-L55)

----
> but how to set Ocelot `HttpConnection`'s Timeout?

The link: [code](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Requester/HttpClientBuilder.cs#L25-L27):
```csharp
            // This is hardcoded at the moment but can easily be added to configuration
            // if required by a user request.
            _defaultTimeout = TimeSpan.FromSeconds(90);
```

----
Are there any other problems we can help you with? 🤣
